### ISSUE TITLE
Make async futures `Unpin` if they don't borrow anything across `.await`s

### ISSUE THREAD

**Issue by jplatte** [rust-lang/rust#82187]
Am I missing something or could rustc in principle make the anonymous (return) type of `async fn` and `async {}` blocks implement `Unpin` (or maybe it should be "not implement `!Unpin`") if the body does not contain any borrows across `.await` points? I couldn't find an existing issue about this.

**Comment by JakobDegen** [comment]
This seems like it would cause problems regarding accidental breaking changes. Changing a function to now borrow something across an await point is an *incredibly* easy change to accidentally make and would result in the `Unpin` impl being removed from the return type.

An alternative would be to make it possible to explicitly mark that an async function/block should be `Unpin` (removing the possibility of an accidental breaking change). I don't see how to do this in function signatures without introducing an additional keyword though (`pub unpin async fn` or such) and that's probably not worth the effort.

**Comment by jplatte** [comment]
@JakobDegen I don't see how `Unpin` would be different to `Send` and `Sync` in this regard.

**Comment by hyd-dev** [comment]
How about doing this *only* for `async` blocks? `async fn`s will remain returning `impl Future` without `+ Unpin`, and function authors that want to make it `Unpin` can specify the return type as `impl Future + Unpin` manually.

**Comment by jplatte** [comment]
> function authors that want to make it `Unpin` can specify the return type as `impl Future + Unpin` manually

You mean by making it a regular `fn` and wrapping the entire body in `async move {}`? That would be pretty annoying. I think a similar strategy could be used to create an attribute macro that asserts `Send` / `Sync` / `Unpin` for `async fn` (not sure about `async` blocks, can proc-macro attrs be used on expressions?), but doing it manually would be super annoying.

**Comment by hyd-dev** [comment]
> You mean by making it a regular `fn` and wrapping the entire body in `async move {}`? That would be pretty annoying. I think a similar strategy could be used to create an attribute macro that asserts `Send` / `Sync` / `Unpin` for `async fn` (not sure about `async` blocks, can proc-macro attrs be used on expressions?), but doing it manually would be super annoying.

Yes, and I agree that could be annoying (sorry for not thinking of this point before commenting, stability itself is annoying)...

**Comment by jplatte** [comment]
I guess I'll create that then 😄

**Comment by jplatte** [comment]
The amount of code required is smaller than I expected. I've hardly tested it, but I think it should work: https://docs.rs/async_auto_traits/0.1

**Comment by hyd-dev** [comment]
Maybe off-topic: @jplatte Thanks for the crate :heart:. I have a question though: is `#[async_auto_traits::clear]` supposed to enforce no auto trait implementations can be used (if it's not, ignore me)? Looks like it does not (this compiles):
```rust
#[async_auto_traits::clear]
async fn foo() {}

fn assert_send(_: impl Send) {}

fn main() {
    assert_send(foo());
}
```
And it looks like `impl Future` [does not hide auto traits](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=580746c7682bc361403a6c002e67ab28) (so I guess that attribute actually does nothing currently and `#[set]` is equivalent to `#[assert]`).

**Comment by jplatte** [comment]
That's kind of surprising but kind of not. I'm out of time for today, testing whether one can explicitly opt out via `+ !Send` or sth. like that would be interesting.

**Comment by hyd-dev** [comment]
Looks like this can prevent the `Future` from implementing `Send` and `Sync` ([play](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=f9d1e1f625b7962fb8624c0948a5c031)):
```rust
use std::{future, marker::PhantomData};

async fn foo() {
    future::ready(PhantomData::<*mut ()>).await;
}

fn assert_send_sync(_: impl Send, _: impl Sync) {}

fn main() {
    assert_send_sync(foo(), foo());
}
```

**Comment by jplatte** [comment]
Okay, let's move this discussion to an issue on https://github.com/jplatte/async_auto_traits.

**Comment by mamins1376** [comment]
> An alternative would be to make it possible to explicitly mark that an async function/block should be `Unpin` (removing the possibility of an accidental breaking change). I don't see how to do this in function signatures without introducing an additional keyword though (`pub unpin async fn` or such) and that's probably not worth the effort.

@JakobDegen You can already assert that the return type of your `async fn` is `Send` on stable ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4dc72d19eb9175124939d722dd235e70)):
```rust
#![allow(dead_code)]

/// Part of crate's API surface.
pub async fn foo() {
    //let _rc = std::rc::Rc::new(0); // uncomment to explode!
    std::future::pending().await
}

// inspired by static_assertions crate
const _: fn() = || {
    fn assert_is_send(_val: impl Send) {}
    assert_is_send(foo());
};
```

Which can easily be automated with a macro. I don't see how `Unpin` would be any different.

*Sidenote*: [Return Type Notation](https://smallcultfollowing.com/babysteps/blog/2023/02/13/return-type-notation-send-bounds-part-2/) is a hot thing right now but seems pretty promising too.

**Comment by Dylan-DPC** [comment]
Closing this as it's better tracked in the https://github.com/jplatte/async_auto_traits/issues/1 issue, and it is better to have it contained in one issue instead of having 2 issues tracking it in different places

**Comment by jplatte** [comment]
@Dylan-DPC I think you misunderstood something, that is about preventing the leak of auto-traits of `async fn` future types. It only relates to `Unpin` in that it would be useful to be able to stop leaking an `Unpin` implementation in a public interface *if the compiler generated those impls at all* (which this issue is about).

**Comment by djc** [comment]
I was experimenting today with replacing the likes of `futures_util::future::ok()` (which is IMO a little ugly -- it reeks of the futures 0.1 times) with `async move {}` blocks and was surprised that I needed an explicit `pin!()` call wrapped around the `async move {}` block even with trivial `Copy` types (like `usize`).

**Comment by kayabaNerve** [comment]
I ran into this, unfortunately. While I completely understand the commentary on it being too prone to accidentally breaking changes, I'd love if if it was explicitly marked, then Unpin would be eligible to propagate up the library.

One note is I don't have Unpin async blocks though. I have async blocks which are naturally Unpin if their generics are Unpin. I accordingly can't guarantee the returned future is Unpin. For my use-case, I'd need that conditional to be modelled.
